### PR TITLE
Partial Rewrite to remove workspaces, add linked projects, and more testing and bug fixes

### PR DESCRIPTION
**Pull request by elcritch**
- Removes distinction between project and workspace (still a few commands to test/fix)
- SemVer fixed and produces simpler SAT formular
- All Git operations use the `-C <dir>` arg avoiding needing to change dir and bugs with that
- Adds CLI options to dump dep graphs as json and dumping the SAT formula, etc
- Git operations all go through `gitops` module, removing duplication of logic and bugs
- Nimble parsing now uses `git show commit:file` and caching to halve load times
- Properly loads nim-lang/packages aliases fixing a lot of name lookups
- Fixes bugs with handling branch lookups (was broken before)
- More correct handling of project commit history
- Improved logging and output defaults
- Fixes bug with forked projects https://github.com/nim-lang/nimble/issues/1329 
- Adds option to let user override conflicting packages with the same module name

### CODE REVIEW

File: src/basic/versions.nim [hunk 1]
```diff
@@ -235,9 +313,12 @@ proc parseTaggedVersions*(outp: string): seq[Commit] =
       while i < line.len and line[i] in Whitespace: inc i
       while i < line.len and line[i] notin Digits: inc i
       let v = parseVersion(line, i)
-      if v != Version(""):
-        result.add Commit(h: line.substr(0, commitEnd-1), v: v)
-  result.sort proc (a, b: Commit): int =
+      let c = initCommitHash(line.substr(0, commitEnd-1), FromGitTag)
+      if c.isEmpty():
+        continue
+      if v != Version("") or not requireVersions:
```
Comment (Araq) [inline src/basic/versions.nim:319]:
```suggestion
      if not c.isEmpty() and (v != Version("") or not requireVersions):
```

File: src/pkgsearch.nim [hunk 2]
```diff
@@ -34,7 +34,7 @@ proc determineCandidates*(pkgList: seq[PackageInfo];
               result[2].add pkg
               break termLoop
 
-proc singleGithubSearch(c: var Reporter; term: string, fullSearch = false): JsonNode =
+proc singleGithubSearch(term: string, fullSearch = false): JsonNode =
```
Comment (Araq) [inline src/pkgsearch.nim:37]: I fail to see the benefit in removing the `Reporter` parameter.

File: src/runners.nim [hunk 3]
```diff
@@ -60,12 +60,12 @@ include $1
 
 """
 
-proc runNimScript*(c: var Reporter; scriptContent: string; name: string) =
+proc runNimScript*(scriptContent: string; name: string) =
```
Comment (Araq) [inline src/runners.nim:63]: Likewise.

File: src/pkgsearch.nim (hunk 2 above)
Comment (elcritch) [reply src/pkgsearch.nim:37]: Eventually it got to where almost every function had a `Reporter` parameter. It also complicates almost every function call with another argument to decipher.

Plus it's easier to be able to just call `debug pkg, "some note"` in smaller functions without plumbing in `Reporter` through possibly a whole chain of helper functions.

Moving Reporter to a global just simplified a bunch. It's not that different than having `stdout` be a global.

Comment (Araq) [reply src/pkgsearch.nim:37]: > Eventually it got to where almost every function had a Reporter parameter.

Well that sucks. Most functions have nothing to report.

The original idea was that it becomes easy to see which procs actually produce output on stdout and which ones don't. I'm not a fan of excessive logging output (it always fills up some harddrive IME) and for debugging `echo` works well enough.

Comment (elcritch) [reply src/pkgsearch.nim:37]: > Well that sucks. Most functions have nothing to report.

Well I ended up needing to debug many of the smaller functions a lot for little edge cases. I made use of trace vs debug vs info pretty extensively for debugging since you can enable or disable it for specific tests. Plain echo doesn't work as well for debugging some of those test cases since they spit out a lot of extra stuff with echo's. 

Part of that was doing a fairly big overhaul on the gitops and url creation. Now those wouldn't really need logging as they're centralized and well unit tested.

I like the ideal, I just got frustrated with it and made them global at some point.

Comment (Araq) [reply src/pkgsearch.nim:37]: > Now those wouldn't really need logging as they're centralized and well unit tested.

Exactly. The code should be clean, not full of debugging leftovers.

Comment (elcritch) [reply src/pkgsearch.nim:37]: > > Now those wouldn't really need logging as they're centralized and well unit tested.
> 
> Exactly. The code should be clean, not full of debugging leftovers.

Fair point, and probably all the `trace` logs can be deleted as they're mostly debugging leftovers. It'd be easier to just remove trace though.

There's still a lot of functions that provide useful logging for debugging workspace issues, not Atlas itself. Many of those cases are in leaf functions which are a pain to plumb `Reporter` out to. 

Actually now I think about all the gitops needed reporter to log git errors not just debugging atlas bugs, etc. Things like a branch missing in a forked repo that users may want to know about so they can fix their stuff. That tends to mean lots of functions needed Reporter. 

Given Atlas's nature of dealing with git repos and external tools it needs to log a fair number of things for people to understand whatever issue they're having. 

> I'm not a fan of excessive logging output (it always fills up some harddrive IME) and for debugging echo works well enough.

That's a good point. Like Nimble somehow produces lots of output but still misses the output needed to figure out many issues.

I figure it's better to just spend time pruning the output. I went through a couple of times to reduce the amount of default logging output for each logging level.

It probably could use a bit more pruning too and cleanup though.

**Comment by elcritch** [comment]
> Well CI is red.

Now it's all green. But there aren't tests for lockfiles, pinning, or most of the other CLI options.

**Comment by elcritch** [comment]
Added:

* initial link setup
* rename to match nimble links
* rename workspace to project
* add example nimble-link
* add link integration test
* add link integration test - windows hackery
* add link integration test
* fix link - cleanup

File: src/atlas.nim [hunk 4]
```diff
@@ -33,37 +33,28 @@ const
...
-                        URL that matches `filter` if a filter is given
-  updateDeps [filter]
-                        update every dependency that has a remote
+  updateDeps [filter]   update every dependency that has a remote
                         URL that matches `filter` if a filter is given
   tag [major|minor|patch]
                         add and push a new tag, input must be one of:
                         ['major'|'minor'|'patch'] or a SemVer tag like ['1.0.3']
-                        or a letter ['a'..'z']: a.b.c.d.e.f.g
+                        or a letter ['a'..'z']: a.b.context().d.e.f.g
```
Comment (Araq) [inline src/atlas.nim:57]:
```suggestion
                        or a letter ['a'..'z']: a.b.c.d.e.f.g
```

File: src/atlas.nim [hunk 5]
```diff
@@ -44,6 +44,8 @@ Command:
   use <url|pkgname>     add package and its dependencies to the project
                         and patch the project's Nimble file
   install               use the nimble file to setup the project's dependencies
+  link <path>           link an existing project into the current project
+                        to share it's dependencies
```
Comment (ZoomRmc) [inline src/atlas.nim:48]:
```suggestion
                        to share its dependencies
```

Comment (elcritch) [reply src/atlas.nim]: Thanks!

**Comment by ZoomRmc** [comment]
How did you come to the decision to use per-project `deps` directory? Why not use a global repo or a workspace? Have nothing in particular against it yet, just would like to understand the big picture.

**Comment by elcritch** [comment]
> How did you come to the decision to use per-project `deps` directory? Why not use a global repo or a workspace? Have nothing in particular against it yet, just would like to understand the big picture.

See https://github.com/nim-lang/atlas/issues/133

**Comment by elcritch** [comment]
Readded `tag`, `update`, and `updateDeps`. Cleaned up the docs a bit more. Also swapped the `activateGraph` to use `checkoutGitCommitFull` which should update git submodules as well.

File: src/basic/context.nim [hunk 6]
```diff
@@ -28,67 +34,93 @@ type
...
+    pluginsFile*: Path
     proxy*: Uri
-    dumbProxy*: bool
 
+var atlasContext = AtlasContext()
 
-proc `==`*(a, b: CfgPath): bool {.borrow.}
+proc setContext*(ctx: AtlasContext) =
+  atlasContext = ctx
+proc context*(): var AtlasContext =
```
Comment (Araq) [inline src/basic/context.nim:73]: That's terrible btw but can be fixed in a followup PR.

File: src/basic/depgraphtypes.nim [hunk 7]
```diff
@@ -1,183 +1,111 @@
...
+    if visited[url] == NotVisited:
+      if not checkCycles(pkg):
+        return false
+  
+  return true
+
+proc toDestDir*(g: DepGraph; d: Package): Path =
   result = d.ondisk
 
-proc enrichVersionsViaExplicitHash*(versions: var seq[DependencyVersion]; x: VersionInterval) =
```
Comment (Araq) [inline src/basic/depgraphtypes.nim:74]: How come this logic isn't required anymore?

Comment (elcritch) [reply src/basic/depgraphtypes.nim:74]: The logic is there, it's now handled in an additional pass at the end of `expandGraph` by calling `traverseDependency` with an `ExplicitVersions` traverse mode. That way all the explicit versions in all requirements are processed.

File: config.nims [hunk 8]
```diff
@@ -30,22 +37,54 @@ task buildRelease, "Build release":
...
+  if not dirExists("atlas-tests"):
+    cleanTestsTask()
+    echo "Downloading Test Repos zip"
+    exec(fmt"curl -L -o {file} {url}")
+    echo "Unzipping Test Repos"
+    exec(fmt"unzip -o {file}")
+  else:
+    let actualver =
+      if fileExists("atlas-tests/atlas_tests.nimble"):
+        readFile("atlas-tests/atlas_tests.nimble").split("=")[^1].replace("\"","").strip()
```
Comment (narimiran) [inline config.nims:62]: This fails on Nim's "Docs CI" with:

```
/home/runner/work/Nim/Nim/pkgs/atlas/config.nims(62, 9) Error: undeclared identifier: 'readFile'
candidates (edit distance, scope distance); see '--spellSuggest': 
 (3, 6): 'rmFile'
/home/runner/work/Nim/Nim/pkgs/atlas/config.nims(48, 1) template/generic instantiation of `task` from here
/home/runner/work/Nim/Nim/pkgs/atlas/config.nims(62, 17) Error: attempting to call routine: 'readFile'
  found 'readFile' [unknown declared in /home/runner/work/Nim/Nim/pkgs/atlas/config.nims(62, 9)]
/home/runner/work/Nim/Nim/pkgs/atlas/config.nims(48, 1) template/generic instantiation of `task` from here
/home/runner/work/Nim/Nim/pkgs/atlas/config.nims(62, 17) Error: attempting to call routine: 'readFile'
  found 'readFile' [unknown declared in /home/runner/work/Nim/Nim/pkgs/atlas/config.nims(62, 9)]
```

Comment (Araq) [reply config.nims:62]: IMO Nim's Docs CI shouldn't build Atlas docs, it's not required.
